spatialdata/_core/aggregate.py

```diff
@@ -0,0 +1,157 @@
+        elif values_type is ShapesModel:
+            return _aggregate_shapes_by_shapes(values, by, id_key, value_key=value_key, agg_func=agg_func)
+        else:
+            raise NotImplementedError(f"Cannot aggregate {values_type} by {by_type}")
+    else:
+        raise NotImplementedError(f"Cannot aggregate {values_type} by {by_type}")
+
+
+def _aggregate_points_by_shapes(
+    points: ddf.DataFrame | pd.DataFrame,
+    shapes: gpd.GeoDataFrame,
+    id_key: str,
+    *,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    # Have to get dims on dask dataframe, can't get from pandas
+    dims = sd.get_dims(points)
+    if isinstance(points, ddf.DataFrame):
+        points = points.compute()
+    points = gpd.GeoDataFrame(points, geometry=gpd.points_from_xy(*[points[dim] for dim in dims]))
+    return _aggregate(points, shapes, id_key, value_key, agg_func)
+
+
+def _aggregate_shapes_by_shapes(
+    values: gpd.GeoDataFrame,
+    by: gpd.GeoDataFrame,
+    id_key: str,
+    *,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    def circles_to_polygons(df: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+        # We should only be buffering points, not polygons. Unfortunately this is an expensive check.
+        values_geotypes = list(values.geom_type.unique())
+        if values_geotypes == ["Point"]:
+            df = df.set_geometry(df.geometry.buffer(df[ShapesModel.RADIUS_KEY]))
+        elif "Point" in values_geotypes:
+            raise TypeError("Geometry contained shapes and polygons.")
+        return df
+
+    values = circles_to_polygons(values)
+    by = circles_to_polygons(by)
+
+    return _aggregate(values, by, id_key, value_key, agg_func)
+
+
+def _aggregate(
+    value: gpd.GeoDataFrame,
+    by: gpd.GeoDataFrame,
+    id_key: str,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    """
+    Aggregate points by polygons.
+
+    Params
+    ------
+    value
+        GeoDataFrame of points to aggregate
+    by
+        GeoDataFrame of polygons
+    id_key
+        Column in values that indicate value type, e.g. probe id or organelle type
+    value_key
+        Column in values that indicate point value, e.g. probe intensity. If not provided, uses a fill value of 1.
+    agg_func
+        Aggregation function to apply over point values, e.g. "mean", "sum", "count".
+        Passed to pandas.DataFrame.groupby.agg.
+    """
+    assert pd.api.types.is_categorical_dtype(value[id_key]), f"{id_key} must be categorical"
+
+    if by.index.name is None:
+        by.index.name = "cell"
+    by_id_key = by.index.name
+
+    joined = by.sjoin(value)
+
+    if value_key is None:
+        point_values = np.broadcast_to(True, joined.shape[0])
+        value_key = "count"
+    else:
+        point_values = joined[value_key]
+
+    to_agg = pd.DataFrame(
```

Comment:
I think it would be nice to factor out the part of this function that generates the `to_agg` DataFrame (i.e., lines 118-132). I think we will likely want this for aggregating tables that annotate the regions.

Reply:
Can you give me some specific examples of use-cases here? I'm not sure I completley understand what you're asking for.

I'm wondering if this can be solved by joining the table here. Or if this isn't something that would be exposed in this PR, if it's something that can be factored out whenever this other feature is developed.